agent/consul/leader_connect_ca_test.go

```diff
@@ -0,0 +1,256 @@
+func (m *mockCAProvider) ActiveRoot() (string, error)           { return m.rootPEM, nil }
+func (m *mockCAProvider) GenerateIntermediateCSR() (string, error) {
+	m.callbackCh <- "provider/GenerateIntermediateCSR"
+	return "", nil
+}
+func (m *mockCAProvider) SetIntermediate(intermediatePEM, rootPEM string) error {
+	m.callbackCh <- "provider/SetIntermediate"
+	return nil
+}
+func (m *mockCAProvider) ActiveIntermediate() (string, error)                       { return m.rootPEM, nil }
+func (m *mockCAProvider) GenerateIntermediate() (string, error)                     { return "", nil }
+func (m *mockCAProvider) Sign(*x509.CertificateRequest) (string, error)             { return "", nil }
+func (m *mockCAProvider) SignIntermediate(*x509.CertificateRequest) (string, error) { return "", nil }
+func (m *mockCAProvider) CrossSignCA(*x509.Certificate) (string, error)             { return "", nil }
+func (m *mockCAProvider) SupportsCrossSigning() (bool, error)                       { return false, nil }
+func (m *mockCAProvider) Cleanup() error                                            { return nil }
+
+func waitForCh(t *testing.T, ch chan string, expected string) {
+	select {
+	case op := <-ch:
+		if op != expected {
+			t.Fatalf("got unexpected op %q, wanted %q", op, expected)
+		}
+	case <-time.After(3 * time.Second):
+		t.Fatalf("never got op %q", expected)
+	}
+}
+
+func waitForEmptyCh(t *testing.T, ch chan string) {
+	select {
+	case op := <-ch:
+		t.Fatalf("got unexpected op %q", op)
+	case <-time.After(1 * time.Second):
+	}
+}
+
+func testCAConfig() *structs.CAConfiguration {
+	return &structs.CAConfiguration{
+		ClusterID: connect.TestClusterID,
+		Provider:  "mock",
+		Config: map[string]interface{}{
+			"LeafCertTTL":         "72h",
+			"IntermediateCertTTL": "2160h",
+		},
+	}
+}
+
+// initTestManager initializes a CAManager with a mockCAServerDelegate, consuming
+// the ops that come through the channels and returning when initialization has finished.
+func initTestManager(t *testing.T, manager *CAManager, delegate *mockCAServerDelegate) {
+	initCh := make(chan struct{})
+	go func() {
+		require.NoError(t, manager.InitializeCA())
+		close(initCh)
+	}()
+	for i := 0; i < 5; i++ {
+		select {
+		case <-delegate.callbackCh:
+		case <-time.After(3 * time.Second):
+			t.Fatal("failed waiting for initialization events")
+		}
+	}
+	select {
+	case <-initCh:
+	case <-time.After(3 * time.Second):
+		t.Fatal("failed waiting for initialization")
+	}
+}
+
+func TestCAManager_Initialize(t *testing.T) {
+	conf := DefaultConfig()
+	conf.ConnectEnabled = true
+	conf.PrimaryDatacenter = "dc1"
+	conf.Datacenter = "dc2"
+	delegate := NewMockCAServerDelegate(t, conf)
+	manager := NewCAManager(delegate, testutil.Logger(t), conf)
+
+	// Call InitializeCA and then confirm the RPCs and provider calls
+	// happen in the expected order.
+	go func() {
+		require.EqualValues(t, CAStateUninitialized, manager.state)
```

Comment:
IIRC it's not usually a good idea to use `testing.T` from goroutines. The alternative would be using a `chan error` to pass the assertion results back up to the body of the test function for assertion.

Also would it make sense to do lines 199 and 201 outside of the goroutine and just use the goroutine to handles the `initializeCA` call?